Spigot-API-Patches/0243-Added-SoundBuilder.patch

```diff
@@ -0,0 +1,331 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Fruxz <cedricspitzer@outlook.de>
+Date: Fri, 18 Dec 2020 01:02:39 +0100
+Subject: [PATCH] Added SoundBuilder
+
+
+diff --git a/src/main/java/io/papermc/paper/sound/SoundBuilder.java b/src/main/java/io/papermc/paper/sound/SoundBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d257f339694c833ed63042bd64b0b740fa51caed
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/sound/SoundBuilder.java
+@@ -0,0 +1,225 @@
++package io.papermc.paper.sound;
++
++import org.bukkit.Location;
++import org.bukkit.Sound;
++import org.bukkit.SoundCategory;
++import org.bukkit.World;
++import org.bukkit.entity.Player;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Collection;
++import java.util.List;
++
++/**
++ * Helps prepare a sound to be sent to players.
++ */
++public interface SoundBuilder {
++
++    /**
++     * Plays the sound to all receiving players (or all). This method is safe to use
++     * Asynchronously
++     *
++     * @return a reference to this object.
++     */
++    SoundBuilder play();
++
++    /**
++     * @return The sound going to be sent
++     */
++    @NotNull
++    Sound sound();
++
++    /**
++     * Changes what sound will be sent
++     *
++     * @param sound The sound
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder sound(@NotNull Sound sound);
++
++    /**
++     * @return The category under which the sound will be played
++     */
++    @Nullable
++    SoundCategory category();
++
++    /**
++     * Changes under which category the sound will be played
++     */
++    @NotNull
++    SoundBuilder category(SoundCategory category);
++
++    /**
++     * @return The volume level of the sound played
++     */
++    @NotNull
++    Float volume();
++
++    /**
++     * Changes the volume level of the sound played
++     *
++     * @param volume The volume
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder volume(@NotNull Float volume);
++
++    /**
++     * @return The pitch of the sound played
++     */
++    @NotNull
++    Float pitch();
++
++    /**
++     * Changes the pitch of the sound played
++     *
++     * @param pitch The pitch
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder pitch(@NotNull Float pitch);
++
++    /**
++     * @return List of players who will receive the sound, or null for all in world
++     */
++    @Nullable
++    List<Player> receivers();
```

Comment:
`List` is an implementation detail, I would suggest `Collection` instead.

Reply:
I disagree with that. Returning list allows using index search operations without having to explicity cast/guess the object as List and distinguishes immediately that the Collection differs from behaviour of another (e.g. a set) in that it allows multiple entries etc.

Reply:
With Paper, such things are also no longer supplied with Collections, this is now always done with Lists.

Reply:
Since a `List` is returned index search is not guaranteed to be efficient, it could be a `LinkedList` after all. The docs never mention that the order of elements is reliable, therefore that is an implementation detail plugins should not depend on. Having the same receiver added multiple times seems like a bug, not a feature.

Reply:
Converting List to Collection just seems ambiguous when there is no reason to. If it was returning an ArrayList, then yes I'd say that is too big of an implementation detail over List.